src/session_impl.cpp

```diff
@@ -1973,6 +1984,143 @@ namespace {
 #endif
 	}
 
+	void session_impl::reopen_outgoing_sockets()
+	{
+		// first build a list of endpoints we should be listening on
+		// we need to remove any unneeded sockets first to avoid the possibility
+		// of a new socket failing to bind due to a conflict with a stale socket
+		std::vector<listen_endpoint_t> eps;
+
+		for (auto const& iface : m_outgoing_interfaces)
+		{
+			interface_to_endpoints(iface, 0, false, eps);
+#ifdef TORRENT_USE_OPENSSL
+			interface_to_endpoints(iface, 0, true, eps);
+#endif
+		}
+
+		// if no outgoing interfaces are specified, create sockets to use
+		// any interface
+		if (eps.empty())
+		{
+			eps.emplace_back(address_v4(), 0, "", false);
```

Comment:
@ssiloti I can't remember anymore why the `outgoing_sockets` must be initialized and used. I can see that the utp_socket_manager can only use sockets from the outgoing_sockets list. It seems to me that it's overlapping a lot with the listen sockets list, especially for UDP.

Me looking at this is related to the SOCKS5 issues people are having, and this [comment/observation](https://github.com/qbittorrent/qBittorrent/issues/11735#issuecomment-568610944) I made.

Even when I set a single listen socket, I get 3 SOCKS5 connections. 1 for the listen socket UDP socket, and 2 for the outgoing socket UDP sockets (because I don't specify any outgoing interface). I would have expected that `m_outgoing_sockets` would be empty, and that the listen socket would be used as the UDP socket for the `utp_socket_manager`.

What would break if `m_outgoing_sockets` were removed entirely, and `utp_socket_manager` was changed to use the UDP sockets from the `m_listen_sockets` array instead?

iirc, the original use of `outgoing_interfaces` was to bind outgoing TCP sockets to specific interfaces. It's not obvious to me what purpose they serve for UDP (that doesn't overlap with listen sockets).

Reply:
IIRC ougoing_sockets is used with UTP because with IPv6 listen sockets are bound to a specific address by default and libtorrent doesn't necessarily know which interface should be used for an outgoing connection. Using the outgoing socket(s), which are bound to the unspecified address by default, allows the OS to decide how to route outgoing connections.

Reply:
I think one important property is that DHT and UDP trackers use the same socket as uTP. since those will open up the NAT pinhole that incoming uTP connections will be received through. as far as I can tell now, utp_socket_manager doesn't use the listen sockets at all, but the DHT and UDP trackers do, so there is a disconnect there